core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala

```diff
@@ -1767,8 +1767,13 @@ private[spark] class DAGScheduler(
 
           // TODO: mark the executor as failed only if there were lots of fetch failures on it
           if (bmAddress != null) {
-            val hostToUnregisterOutputs = if (env.blockManager.externalShuffleServiceEnabled &&
-              unRegisterOutputOnHostOnFetchFailure) {
+            val externalShuffleServiceEnabled = env.blockManager.externalShuffleServiceEnabled
+            val isHostDecommissioned = taskScheduler
+              .getExecutorDecommissionInfo(bmAddress.executorId)
+              .exists(_.isHostDecommissioned)
```

Comment:
So would we want to do this unless local fetch is disabled (which since it isn't merged yet it will be) and the executor is decommissioned or the the host is decommissioned? Should we use isShuffleLost here instead?

Reply:
@holdenk .. by "local fetch", did you mean https://github.com/apache/spark/pull/28911 ? 

Typically `isHostDecommissioned` is set by the cluster manager: It knows for sure that the host is lost or not. For example, the Master knows when a Worker goes away for real. Something similar would have to be done for YARN.

Is the reasoning for adding this check: We don't need to clear the shuffle state if an executor is decommissioned even if external shuffle service is disabled, as long as there are other executors on that host ? 

If I understand you correctly, the condition you are hinting at is this (please do confirm):

```
   if (host is decommissioned) {
    // clear shuffle state for the entire host
   } else if (!isLocalFetchEnabled || numRemainingLiveExecutorsOnThisHost == 0) {
    // clear shuffle state for just this executor that suffered the fetch failure
    // because we can be sure that no one else will serve this executor's data
   }
```

I will consider using `isShuffleLost ` here to determine if the shuffle has been lost here.

Reply:
`isShuffleLost` is not very applicable here. Actually that method was too narrow in scope so I inlined it. It strictly means that shuffle is lost for this executor. 

In this context, we already know that shuffle is lost for the executor: We are simply trying to determine if it is also lost for the entire host. I updated the code to reflect the logic/intent better. 

I will create a follow up Jira under the master ticket to track changing this logic when "Local Fetch" is merged in.

Reply:
Yes that is correct reading of my comment. In the else if block though I'd also add that the external shuffle service is not enabled.

Reply:
I thought a bit more about this and I think that the author of #28911 should implement the optimization of whether or not a shuffle state should be cleared when you get a fetchf failure from an executor -- when other "local" executors might serve those outputs. 

This is an optimization in that it leverages the Local fetch feature introduced in #28911. I don't know enough about local fetch implementation to comment more about it.

From my perspective of decommissioning, I have changed the logic to be: Executors on the host are fate shared if either we know that they share an external shuffle service or if we know that the host has been decommissioned. In that case, mark the entire host as lost (if the feature flag `unRegisterOutputOnHostOnFetchFailure` is enabled).

cc: @Ngone51 (author of #28911)

Reply:
Sounds reasonable. So long as we reconcile how these two features work together before 3.1.

Reply:
> if (host is decommissioned) {
    // clear shuffle state for the entire host
   } else if (!isLocalFetchEnabled || numRemainingLiveExecutorsOnThisHost == 0) {
    // clear shuffle state for just this executor that suffered the fetch failure
    // because we can be sure that no one else will serve this executor's data
   }

> should implement the optimization of whether or not a shuffle state should be cleared when you get a fetchf failure from an executor 

These're actually good idea! I just have a few concerns that whether they are doable with current implementation... I mean, when fetch failure happens on an executor, the possible reasons may be executor lost or shuffle data corrupt. For these two reasons, host-local shuffle reading could still result in fetch failure at the end. Because, even if we use host-local shuffle data reading, we need to ask the lost executor for the directory first. And if the shuffle data is corrupted, we'll fail to read the shuffle data anyway. Though, in the case of executor lost, it's maybe doable as we have local disk directories cache locally so that the executor on the same host won't ask the lost executor for the directory. However, the problem is that the driver is not aware of the status of the cache. Thus, the driver also has no idea whether we should clean up the shuffle state for the lost executor or not.